### ISSUE TITLE
include costs for ThresholdOptimizer

### ISSUE THREAD

**Issue by deltasun** [fairlearn/fairlearn#473]
I suggest that the constructor `ThresholdOptimizer()` include the possibility of considering asymmetric cost in False Positives and False Negatives. 

As by actual implementation, thresholds are chosen such that given the constraints, the results minimize `P(Yhat = 1, Y = 0) + P(Yhat = 0, Y = 1)`, that is the total fraction of false positives and false negatives (which is equivalent to maximize accuracy).
I simply suggest to add `Cfp * P(Yhat = 1, Y = 0) + Cfn * P(Yhat = 0, Y = 1)`, with default `Cfp=Cfn=1`; or even a single parameter (corresponding e.g. to Cfp/Cfn.

The most general case would be to compute the error as:
`Ctp * P(Yhat = 1, Y = 1) + Ctn * P(Yhat = 0, Y = 0) + Cfp * P(Yhat = 1, Y = 0) + Cfn * P(Yhat = 0, Y = 1)` but it is pretty reasonable to assume that both true positives and true negatives have no cost at all.

By the way, this is also coherent with Hardt et al. (2016) which is the main reference for `ThresholdOptimizer()`.

This issue may be related to PR https://github.com/fairlearn/fairlearn/pull/381, but I cannot see references to cost modification in the code of this PR. 

The changes should be minimal, just add these two parameters in the class `__init__` and then use them in the definition of the error

What do you think?

**Comment by MiroDudik** [comment]
I want to get #381 in first, but once that's in, adding the costs will only require minor refactoring. The API in #381 is of the form
* `thr_optimizer = ThresholdOptimizer(constraints=..., objective='accuracy_score', ...)`

We also allow objective to be `'balanced_accuracy_score'` and a few other string options.

For the scenario you are thinking about, I can see two API styles that would match up with this incoming PR:

1. `thr_optimizer = ThresholdOptimizer(constraints=..., objective_cost_matrix=[[0,1],[1,0]], ...)`

   where the cost matrix follows the same convention as sklearn's [confusion_matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html)

2. `thr_optimizer = ThresholdOptimizer(constraints=..., objective_costs = {'fp': 1.0, 'fn': 0.8})`

I think I much prefer the second one for clarity.

**Comment by deltasun** [comment]
Yes @MiroDudik I prefer the second one too! That's exactly what I was thinking about.

I actually tweaked the class just to include 2 extra parameters Cfp and Cfn, but with the dictionary is more elegant

thanks a lot!

(of course if I can be of any help in the code writing as well, let me know. I am not an expert at all in github contributing, but I can learn!)
